Changes to replace scx agent with omsagent

vimish opened the pull request:
@Microsoft/omsagent-devs 
@jeffaco 

These changes allows scx agent to be replaced by omsagent on Linux platforms so that it can be used to enhance log file monitoring of SCOM XPlat monitoring.

Description:
1. Do not disable OMI port 1270 so that agent machine can be discovered by SCOM.
2. Register a secondary workspace for SCOM on the agent machine. The Fluentd process of this secondary workspace will be managed through SCOM.
3. 5 Fluentd filter plugins has been added that will generate events to be send to SCOM server.
4. An output plugin(similar to out_oms plugin) has been added to send the data to SCOM server's endpoint. 

Tests done:
1. PBUILD has been run successfully.
2. Following scenarios are tested:
        i. Only OMS (Primary workspace)
       ii. First OMS(Primary workspace) then SCOM
      iii. First SCOM then OMS(Primary workspace)
      iv. Only SCOM

jeffaco commented:
Just to make sure I understand this properly:

You're not really replacing the SCOM agent, as the commit message (`Changes to replace scx agent with omsagent`) says. You're actually adding logging components. Right? After all, OMS-Agent itself uses SCOM for performance monitoring and other information.

The SCOM agent still exists, and is still being used. Indeed, nothing is being done to keep the SCOM agent from using it's own log file management. Is this all correct? If so, please use a better commit message!

When is OMS's logging components used with SCOM (on Windows) vs. the built-in log file monitor in the opsmgr Linux agent? Can you clarify that behavior?

Next, some of the description (for this PR) is totally useful, and should likely be in the commit message too (use a multi-lined commit). A multi-lined commit that gives useful information about a change is totally useful, and makes it much easier to get background on what was done and why.

For the Ruby changes, I'll leave that to someone else that understands that stuff better. However, I did have some code comments, which I'll get to now.

Regardless, I'm going to mark this "waiting"; I'd like to understand the scope of changes here and how it interacts with SCOM (Windows) vs. the existing log file monitor before signing off. I didn't see any of the Windows components changed, so it seems like this is not replacing logging at all, but adding a new logging mechanism. And near as I can tell, SCOM on Windows doesn't (directly) use it.

jeffaco commented on installer/scripts/omsadmin.sh:
```diff
@@ -310,6 +310,28 @@ set_proxy_setting()
+    configure_logrotate
+
+    # The scx certificate and key needs to be owned by omsagent so that it is readable and can be used during SCOM authentication
+    if [ -f /etc/opt/microsoft/scx/ssl/scx.pem ]; then
```
I'm not quite sure why you're checking for a certificate in that location (at least directly). Isn't this historical for older agents (when OMI and SCX were shipping together)?

Since OMS always installs a newer version of SCX, then I'd think that SCX couldn't possibly have a key stored in that location anymore. Why bother checking for that here? When would this ever be necessary?

jeffaco commented on installer/scripts/omsadmin.sh:
```diff
@@ -310,6 +310,28 @@ set_proxy_setting()
+
+    if [ -f /etc/opt/omi/ssl/omikey.pem ]; then
+       chown_omsagent /etc/opt/omi/ssl/omikey.pem
+    fi
```
The rest of the changes are all Ruby stuff, and I don't know Ruby very well. I'll leave the Ruby review to someone else on the OMS team, or perhaps to @niroyb if he's so inclined to review it.

jeffaco commented on installer/datafiles/linux.data:
```diff
@@ -102,6 +102,7 @@ ${{OMS_SERVICE}} start
 # If we're called for upgrade, don't do anything
 if ${{PERFORMING_UPGRADE_NOT}}; then
     ${{OMS_SERVICE}} disable
+    ${{OMS_SERVICE}} disable scom
```
I see that service_control now takes workspaces (essentially separate services), so disregard this comment.

vimish commented:
@jeffaco 
@Microsoft/omsagent-devs 

Please review the new iteration.
Changes made are:

1. Updated the commit message
2. A new certificate will be created using scxsslconfig tool to be used for this feature. Corresponding [PR](https://github.com/Microsoft/SCXcore/pull/68) for scxsslconfig changes to support this.
3. Added a utility to regenerate this cert. As of now this utility only accepts hostname and domain name as input and keeps other parameters at default. If required will add support for other parameters in future.
4. Removed the changes to enable port 1270 by default. Will open a new PR for it that will add a flag in shell bundle to signify whether to enable port or not.

jeffaco commented on installer/scripts/omsadmin.sh:
```diff
@@ -310,6 +313,36 @@ set_proxy_setting()
+        mv $CERT_DIR/omi-host-*.pem $CERT_DIR/scom-cert.pem
+        mv $CERT_DIR/omikey.pem $CERT_DIR/scom-key.pem
+        rm -f $CERT_DIR/omi.pem
+        chown_omsagent $CERT_DIR/scom-cert.pem
```
Is 333 and 334 necessary? Oh, I guess if you're running as root, they are. But not if you're running as omsagent. But in that case, it's harmless.

lagalbra commented on source/code/plugins/out_scom.rb:
```diff
@@ -0,0 +1,89 @@
+      else
+        raise 'Inavlid Record: #{record}'
+      end
+      @log.info "END: #{scom_endpoint}" 
```
Unless you want the user to be aware of the endpoint, I would recommend using log.debug instead for this.

lagalbra commented on source/code/plugins/out_scom.rb:
```diff
@@ -0,0 +1,89 @@
+      elsif record.has_key? 'PerformanceDataList'
+        scom_endpoint = SCOM::Configuration.scom_perf_endpoint
+      else
+        raise 'Inavlid Record: #{record}'
```
spelling: Invalid

lagalbra commented on source/code/plugins/scom_common.rb:
```diff
@@ -0,0 +1,116 @@
+          res_summary = "class=#{res.class.name}; code=#{res.code}; message=#{res.message}; body=#{res.body};)"
+          raise RetryRequestException, "HTTP error: #{res_summary}"
+        end
+
```
A lot of these methods are copied from [oms_common.rb](https://github.com/Microsoft/OMS-Agent-for-Linux/blob/master/source/code/plugins/oms_common.rb). I would prefer you extract the functionality you're trying to use into helper methods in the oms_common.rb file and then call those.

lagalbra commented on source/code/plugins/scom_configuration.rb:
```diff
@@ -0,0 +1,100 @@
+    require_relative 'omslog'
+        
+    @@configuration_loaded = false
+    @@scom_conf_path = '/etc/opt/microsoft/omsagent/scom/conf/omsadmin.conf'
```
Remember: if you change the workspace-id-placeholder in omsadmin.sh from "scom" to "SCOM", you will have to rename these paths.

lagalbra commented on source/code/plugins/scom_configuration.rb:
```diff
@@ -0,0 +1,100 @@
+        return false
+      elsif monitoring_id_lines.size == 0
+        OMS::Log.warn_once("Found more than one MONITORING_ID setting in #{@@scom_conf_path}, will use the first one.")
+      end
```
This block is the same (except for the value being searched for) as lines 28-34. Please extract into a helper method.

lagalbra commented on test/code/plugins/omstestlib.rb:
```diff
@@ -10,8 +10,10 @@ def initialize
-
-    def trace(message)
+    
+    #Making message optional here because a Fluentd call was throwing ArgunmentError
```
spelling: ArgumentError

lagalbra commented on test/code/plugins/out_scom_test.rb:
```diff
@@ -0,0 +1,69 @@
+
+  class SCOM::Configuration
+    class << self
+      def configuration_Loaded=(configLoaded)
```
nit: for proper casing, change to "def configuration_loaded" (lowercase L)

lagalbra commented on test/code/plugins/out_scom_test.rb:
```diff
@@ -0,0 +1,69 @@
+
+  def test_batch_data
+    # mock the configuration to be loaded
+    SCOM::Configuration.configuration_Loaded = true
```
nit: for proper casing, change to "def configuration_loaded" (lowercase L)

lagalbra commented:
After reviewing design doc and conf files provided through email, I have a couple extra notes/revisions:
1. The design of 5 separate filter plugins seems better from the customer perspective, so don't feel obligated to change that. However, lots of your code should be extracted into your scom_common.rb library.
2. If you want to follow the MongoDB logs, I recommend using our [MongoDB source configuration](https://github.com/Microsoft/OMS-Agent-for-Linux/blob/master/installer/conf/omsagent.d/mongo.conf) and instructing the customer to simply provide a filter configuration block at the bottom of that file to reference your filter plugin. (instead of copying the configuration lines from our file and changing the tag)

KrisBash commented:
These changes:

- Updated the commit message
- A new certificate will be created using scxsslconfig tool to be used for this feature. Corresponding PR for scxsslconfig changes to support this.

should be behind an installation and omsadmin.sh switch. We shouldn't open the TCP port unless the agent is expected to be used with OpsMgr. For OpsMgr cases, the installation can be done with a switch like --OpsMgr

vimish commented on installer/scripts/omsadmin.sh:
```diff
@@ -310,6 +313,36 @@ set_proxy_setting()
+onboard_scom()
+{
+    echo "Onboarding SCOM"
+    create_workspace_directories "scom"
```
As of now updation of config file is manual for us . This will require navigation to the config file location. We think that having small case scom in the path will be easier for navigation hence we prefer to leave it this way.

vimish commented on source/code/plugins/filter_scom_excl_correlation.rb:
```diff
@@ -0,0 +1,73 @@
+module Fluent
```
I will like to open a new PR to extract out common code in the filter plugins after merging this PR. This will give us a base agent for our functionality which we can use to improve further.

vimish commented:
@lagalbra 
Please review the new upload.

I have addressed most of the review comments and added comment for the remaining ones.
Few other points:
1. I have tested it with multihoming feature. SCOM workspace is always registered as a secondary workspace. A basic sanity test is done for following scenarios: 
      i. Only OMS (Primary workspace)
     ii. First OMS(Primary workspace) then SCOM
    iii. First SCOM then OMS(Primary workspace)
    iv. Only SCOM
2. As of now we do not make any assumptions wrt input logs. MongoDB logs were just used for demo purpose. Thanks for sharing the information about MongoDB. We will keep use that in our recommendation to the customers.

vimish commented:
@KrisBash 

1.The registration of SCOM workspace and generation of certificates is done via omsadmin.sh switch. We have updated out MP's to call this during discovery process.
2. I have removed the change to enable port 1270 by default. As suggested by you and Jeff I will be adding a new flag in the shell bundle to determine whether to open the port or not. I will open another PR for this functionality.
